### ISSUE TITLE
Externalize ImageSharp.Drawing

### ISSUE THREAD

**Issue by antonfirsov** [SixLabors/ImageSharp#1002]
Prerequisite: #967

----

**UPDATE**

### Final repository & package structure
![image](https://user-images.githubusercontent.com/6835152/68545258-97e78b80-03cb-11ea-87fc-864be7939dc7.png)

### Refactor steps
I think the best would be to follow the order I suggest. Please let me know if something is missing!

- [ ] Copy all contents of *SixLabors.Core* into the *SixLabors.ImageSharp* repository (test and product code)
- [x] Create a new, 100% self-contained repository for ImageSharp.Drawing, copy the following content into the repository:
    - Current Drawing code:
        - Product code
        - Test code
        - Benchmark code (!!!)
    - Test images used by Drawing tests
    - Reference images used by Drawing tests
    - Product and test code of `SixLabors.Shapes`
- [x] Configure CI for the ImageSharp.Drawing repository
- [ ] Update *ImageSharp.Web* (dogfeeding)

The steps above will also touch #868. If we want to completely resolve that issue, we shall also:
- [ ] Clean-up history of binary artifacts in *SixLabors.ImageSharp* master
- [ ] Copy all reference images to *SixLabors.ImageSharp*

We shall consider introducing GitHub workflows while implementing these steps, but I need more help if we want to do so.

**Comment by antonfirsov** [comment]
I have two competing proposals for reorganizing our GitHub repositories. (Ignoring ImageSharp.Web, since it will be kept as-is.)

### A, 2 Repositories 2 NuGet packages

- SixLabors.ImageSharp
  - Fully internalizing SixLabors.Core
  - Excluding Drawing

- New repository SixLabors.ImageSharp.Drawing
  - Including SixLabors.Fonts, SixLabors.Shapes and Drawing
  - Continue current line of NuGet packages (Next version: beta0008)

### B, 2 Repositories 5 NuGet packages

- Repo: SixLabors.ImageSharp
  - Including SixLabors.Core
    - But define a separate package for it
    - Versions are bound together (Next version: RC1)
  - Excluding Drawing

- Repo: SixLabors.ImageSharp.Drawing
  - Including SixLabors.Fonts, SixLabors.Shapes and Drawing
  - But keep separate NuGet packages
  - Versions are bound together. (Next version: beta0010, also for Drawing)

/cc @tocsoft @dlemstra @JimBobSquarePants

**Comment by JimBobSquarePants** [comment]
I vote for A to keep maintenance costs down.

**Comment by tocsoft** [comment]
I feel SixLabors.Fonts should still be kept a separate repo there is definitely scope there to be a much more useful project standalone from the rest of the projects. 

However I feel shapes and drawing should be merged into a single project/repo and just label it SixLabors.ImageSharp.Drawing. 

We should inline core into imagesharp and just expose them so that drawing and shapes can depend on them directly.

So in conclusion we would end up with 4 packages & repositories.

1. SixLabors.ImageSharp - includes base IS code & codecs + everything applicable from Core.
2. SixLabors.ImageSharp.Drawing - includes all current drawing & shapes code.
3. SixLabors.ImageSharp.Web - includes all current web stuff (unchanged)
4. SixLabors.Fonts - current Fonts library (unchanged).

Each will version independently an inline with changes to dependent projects. i.e. if either fonts of imagesharp is updated we would try to release an updated drawing package using the latest bits.

**Comment by antonfirsov** [comment]
@tocsoft your point is probably valid, since SixLabors.Fonts has [more downloads](https://www.nuget.org/packages/SixLabors.Fonts/1.0.0-beta0009) than [SixLabors.Shapes.Text](https://www.nuget.org/packages/SixLabors.Shapes.Text/1.0.0-beta0009) or [SixLabors.ImageSharp.Drawing](https://www.nuget.org/packages/SixLabors.ImageSharp.Drawing/1.0.0-beta0007). However, a "standalone" font library depending on the whole heavyweight ImageSharp package seems rather strange to me. If we really want to keep it really separate, we should probably avoid that.

How about keeping SixLabors.Core as a separate NuGet package versioned together with ImageSharp? I don't think it's that much extra maintenance.

**Comment by tocsoft** [comment]
Fonts doesn't have any dependencies on ImageSharp... there might possibly be currently a couple of dependencies on some of the point primitives but we can just switch them out to different classes or just Vector2 directly.

**Comment by tocsoft** [comment]
The internals of Fonts definitely have no dependencies on core anyway. Like i said we can just switch the public API to return Vector2.  

However I am not opposed to retaining SixLabors.Core as a package inside the ImageSharp repo and versioned inline with the main ImageSharp package if it makes the public apis & interop between fonts and drawing/core cleaner.

**Comment by antonfirsov** [comment]
It also uses `RectangleF`, but I can't see anything more. So we can remove the dependency by defining a (not so nice) `FontRectangle` struct or similar.

Are font internals memory intensive, or not at all? (=> Is custom memory management a possible optimization enabler?)

**Comment by tocsoft** [comment]
Its been a while since I've been diving deeply into fonts internals but I doubt the memory manager would give us much benefit and in fact would probably end up complicating stuff making maintenance more difficult so I think memory management should probably just be left in ImageSharp and left out of core.

**Comment by antonfirsov** [comment]
Okkay, how about this (let's call it **option Z**):
- Keep Core as a separate repository, but move Memory Management back to ImageSharp
- Keep Fonts as is
- Merge Shapes into the new Drawing repository
  - Note: Shapes already depends on Memory Management, since shapes could be heavy (eg. a tessellated Glyph)
![image](https://user-images.githubusercontent.com/6835152/67530718-48f9e080-f6c0-11e9-9351-ec8f9ac7e816.png)

Benefits:
- Less effort spent on restructuring
- No multi-package repositories
- Still maintainable
- Enables keeping Core and Fonts under Apache license, if we want

**Comment by JimBobSquarePants** [comment]
Ok, I've had a dig around in the repositories and that plan makes sense. Fonts is not memory intensive and does not require memory management.

**Comment by dlemstra** [comment]
Do we really need the Core package in option Z? And I do understand that it would make sense to have a separate Font package but I wonder if people would use it without the Drawing library. Having two libraries would simplify things.

**Comment by JimBobSquarePants** [comment]
It's there so Fonts can reuse the primitives.

**Comment by dlemstra** [comment]
Will need to take a good look at the code but a tiny bit of duplication might not be bad here.

**Comment by tocsoft** [comment]
I think I agree with Dirk lets just drop the Core package and duplicate the 1/2 small structs what we need for Fonts it will simplify the dependency tree.

**Comment by JimBobSquarePants** [comment]
Are there any public API crossover points? I’m thinking handling conversion between structs.

**Comment by tocsoft** [comment]
don't think so, not that can't be replaced by switching usages of `PointF` to `Vector2` which already has implicit conversion to `PointF`

**Comment by antonfirsov** [comment]
@dlemstra I also very curious about those independent use cases of SixLabors.Fonts. But based on NuGet stats, it's clear that we have such users.

**Comment by dlemstra** [comment]
@antonfirsov When you go to https://www.nuget.org/packages/SixLabors.Fonts/1.0.0-beta0009 there is a `GitHub Usage` block and that shows only 2 repos.

**Comment by antonfirsov** [comment]
@dlemstra those are only OSS consumers, and only top repositories are listed.

What I mean is that SixLabors.Fonts has [more downloads](https://www.nuget.org/packages/SixLabors.Fonts/1.0.0-beta0009) than [SixLabors.Shapes.Text](https://www.nuget.org/packages/SixLabors.Shapes.Text/1.0.0-beta0009) or [SixLabors.ImageSharp.Drawing](https://www.nuget.org/packages/SixLabors.ImageSharp.Drawing/1.0.0-beta0007). I see no other explanation for this, than the existence of independent (closed source) use-cases.

**Comment by antonfirsov** [comment]
Posted an update in the issue description summarizing the final plans and the steps to achieve them.
